File: import_build/src/Computation/DataList.py
# // a class for working with points in the configuration space of a
# //composite system.
# //instances of the DataList class are arrays of objects
# //these objects need to implement the methods .clone(), .add(), .sub(), and .multiplyScalar()
# //thus, DataLists can be lists of either states, or dStates as required:
import numpy as np

class DataList:
    """
    A class used to store information about collection of states (collection of vertices)

    ...

    Attributes
    ----------
    data : array
        array of state objects (vertices)

    connectivity : array
        array of connectivity information between states in data - defaults to None i.e. data is collection of free vertices
        (e.g. connectivity = [[1,0],[3,8]] means states 0 and 1 are connected and states 3 and 8 are connected via coupling potential or constraint)

    Methods
    -------
    clone()
        Generate copy of self
        Returns DataList clone

    combine(dataList)
        Generates new DataList object consisting of states from self and dataList 
        Returns combined Datalist object

    toArray()
        Generate copy of self.data as an np.array
        Returns flattened np.array (e.g. for DataList of States np.array([state1.pos, state1.vel, ..., stateN.pos, stateN.vel]).flatten() or for DataList of dStates np.array([state1.vel, state1.acc, ..., stateN.vel, stateN.acc]).flatten())

    add(dataList)
        Add velocity of each state in dataList to each state in self componentwise
        Returns self

    sub(dataList)
        Subtract velocity of each state in dataList from each state in self componentwise
        Returns self

    multiplyScalar(k)
        Scale each state in self by scalar value k
        Returns self

    flow(eps)
        Move each state in self infinitesimally (i.e. by eps) along its tangent vector
        Returns self

    updateBy(dataList)
        Update each state in self by infinitesimally flowing along its respective differential (dState) contained in dataList (DataList of dStates)
        Returns self
    """

    # Populate list with states
    def __init__(self, data, connectivity = []):
        # data is list of State objects (describing motion of vertices)
        self.data = data
        self.connectivity = connectivity
        # self.rig_connectivity = rig_connectivity

        # //implementation of .clone() for the array
    def clone(self):
        temparr = []
        for a in self.data:
            temparr.append(a.clone())
        return  DataList(temparr, self.connectivity)
    
    def combine(self, dataList):
        tempsarr = []
        tempcarr = []

        # Combine state data
        for a in self.data:
            tempsarr.append(a.clone())
        for b in dataList.data:
            tempsarr.append(b.clone())

        # Combine connectivity data
        if len(self.connectivity) != 0:
            tempcarr = tempcarr + self.connectivity
        if len(dataList.connectivity) != 0:
                for c in dataList.connectivity:
                    c = [c[0] + len(self.data), c[1] + len(self.data)]
                    tempcarr.append(c)
        return DataList(tempsarr,tempcarr)
    
    def toArray(self):
        res_array = []

        # For DataList of States
        if self.data[0].__class__.__name__ == "State":
            for c in self.data:
                res_array.append(c.pos.copy())
                res_array.append(c.vel.copy())

        # For DataList of dStates
        if self.data[0].__class__.__name__ == "dState":
            for c in self.data:
                res_array.append(c.vel.copy())
                res_array.append(c.acc.copy())

        res_array = np.array(res_array).flatten()

        return res_array
        

    # //implementing .add() componentwise (between datalists)
    def add(self, dataList ):
        for a in range(len(self.data)):
            # print(a)
            self.data[a] = self.data[a].add(dataList.data[a])
        return self

    # //implementing .sub() componentiwse (between datalists)
    def sub(self, dataList ):
        for a in range(len(self.data)):
            # print(a)
            self.data[a] = self.data[a].sub(dataList.data[a])
        return self
    
    # //implementing .multiplyScalar() componentwise
    def multiplyScalar(self, k ):
        for a in self.data:
            a = a.multiplyScalar(k)
        return self

    def flow(self, eps):
        for a in self.data:
            a = a.flow(eps)
        return self

    # //implementing .updateBy() componentwise
    # //WARNING!!! RIGHT NOW NOT ALL OBJECTS HAVE AN UPDATEBY
    # //ONLY STATE, AND IT TAKES DSTATE: SO THIS WILL RETURN
    # //AN ERROR IF ITS CALLED IN A CASE WHERE THE INDIVIDUAL OBJECTS DONT
    # //IMPLEMENT IT:  THAT'S EXPECTED; BUT WE SHOULD INCLUDE AN ACTUAL ERROR
    # //MESSAGE TO CONSOLE.LOG() THAT EXPLAINS.

    def updateBy(self, dataList ):
        for a in range(len(self.data)):
            # print(a)
            self.data[a] = self.data[a].clone().updateBy(dataList.data[a])
            # print(a)
            # print(self.data[a].pos)
        return self

File: cpp_build/Computation/DataList.cpp
#include <vector>
#include <iostream>

#include "State.h"
#include "DState.h"
#include "DataList.h"
#include "pairwise_connections.h"



// Populate datalist with states and mesh connectivity data
template <typename T>
DataList<T>::DataList()
{

}

template <typename T>
DataList<T>::DataList(std::vector<T> data, std::vector<Pairwise_Connection> connectivity, std::vector<Pairwise_Rigid_Connection> rig_connectivity)
{
    this->data = data;
    this->connectivity = connectivity;
    this->rig_connectivity = rig_connectivity;
}

//make a copy of a given datalist (not just reference it in memory) and return the copy
template <typename T>
DataList<T> DataList<T>::clone()
{
    return  DataList(this->data,this->connectivity,this->rig_connectivity);
}

template <typename T>
DataList<T> DataList<T>::combine(DataList<T> datalist)
{
    std::vector<T> tempsarr;
    std::vector<Pairwise_Connection> tempcarr;
    std::vector<Pairwise_Rigid_Connection> temprarr;

    // Combine state data
    for (int i = 0; i < this->data.size(); i++)
    {
        tempsarr.push_back(this->data[i]);
    }
    for (int i = 0; i < datalist.data.size(); i++)
    {
        tempsarr.push_back(datalist.data[i]);
    }

    // Combine connectivity data
    if (this->connectivity.size() != 0)
    {
        for (int i = 0; i < this->connectivity.size(); i++)
        {
            tempcarr.push_back(this->connectivity[i]);
        }
    }
    if (datalist.connectivity.size() != 0)
    {
        for (int i = 0; i < datalist.connectivity.size(); i++)
        {
            Pairwise_Connection temp_pair;
            // Base vertices for relabeling when adding new set of vertices
            int vert_num = this->data.size();
            temp_pair.index_pair[0] = datalist.connectivity[i].index_pair[0] + vert_num;
            temp_pair.index_pair[1] = datalist.connectivity[i].index_pair[1] + vert_num;
            tempcarr.push_back(temp_pair);
        }
    }

    // Combine rig_connectivity data
    if (this->rig_connectivity.size() != 0)
    {
        for (int i = 0; i < this->rig_connectivity.size(); i++)
        {
            temprarr.push_back(this->rig_connectivity[i]);
        }
    }
    if (datalist.rig_connectivity.size() != 0)
    {
        for (int i = 0; i < datalist.rig_connectivity.size(); i++)
        {
            Pairwise_Rigid_Connection temp_pair;
            // Base vertices for relabeling when adding new set of vertices
            int vert_num = this->data.size();
            temp_pair.index_pair[0] = datalist.rig_connectivity[i].index_pair[0] + vert_num; 
            temp_pair.index_pair[1] = datalist.rig_connectivity[i].index_pair[1] + vert_num;
            temp_pair.lagrange_multipler = datalist.rig_connectivity[i].lagrange_multipler;
            temprarr.push_back(temp_pair);
        }
    }
    return DataList<T>(tempsarr,tempcarr,temprarr);
    
}

template <>
std::vector<double> DataList<State>::toArray()
{
    std::vector<double> temparr;

    for (int i = 0; i < this->data.size(); i++)
    {
        temparr.insert(std::end(temparr), std::begin(this->data[i].pos), std::end(this->data[i].pos));
        temparr.insert(std::end(temparr), std::begin(this->data[i].vel), std::end(this->data[i].vel));
    }

    return temparr;

}

template <>
std::vector<double> DataList<DState>::toArray()
{
    std::vector<double> temparr;

    for (int i = 0; i < this->data.size(); i++)
    {
        temparr.insert(std::end(temparr), std::begin(this->data[i].vel), std::end(this->data[i].vel));
        temparr.insert(std::end(temparr), std::begin(this->data[i].acc), std::end(this->data[i].acc));
    }

    return temparr;

}

//print information about the state to the terminal
template<>
void DataList<State>::print_info()
{
    int count = 0;
    // Add the elements of state.vec and this->vel
    for (State item : this->data)  
    {  
        std::cout << "State " << count << " :\n";
        item.print_info();  
        count++;
    };
}

template<>
void DataList<DState>::print_info()
{
    int count = 0;
    // Add the elements of state.vec and this->vel
    for (DState item : this->data)  
    {  
        std::cout << "DState " << count << " :\n";
        item.print_info();  
        count++;
    };
}

template <typename T>
void DataList<T>::add(DataList<T> datalist)
{
    // Add the elements of state.vec and this->vel
    for (int i = 0; i < this->data.size(); i++)  
    {  
        T temp = datalist.data[i];
        this->data[i].add(temp);  
    };
}

template <typename T>
void DataList<T>::sub(DataList<T> datalist)
{
    // Add the elements of state.vec and this->vel
    for (int i = 0; i < this->data.size(); i++)  
    {  
        T temp = datalist.data[i];
        this->data[i].sub(temp);  
    };
}

template <typename T>
void DataList<T>::multiplyScalar(double k)
{
    // Add the elements of state.vec and this->vel
    for (int i = 0; i < this->data.size(); i++)  
    {  
        this->data[i].multiplyScalar(k);  
    };
}

template <>
void DataList<State>::flow(double eps)
{
    // Add the elements of state.vec and this->vel
    for (State item : this->data)  
    {  
        item.flow(eps);  
    };
}

    // # //implementing .updateBy() componentwise
    // # //WARNING!!! RIGHT NOW NOT ALL OBJECTS HAVE AN UPDATEBY
    // # //ONLY STATE, AND IT TAKES DSTATE: SO THIS WILL RETURN
    // # //AN ERROR IF ITS CALLED IN A CASE WHERE THE INDIVIDUAL OBJECTS DONT
    // # //IMPLEMENT IT:  THAT'S EXPECTED; BUT WE SHOULD INCLUDE AN ACTUAL ERROR
    // # //MESSAGE TO CONSOLE.LOG() THAT EXPLAINS.


template <>
void DataList<State>::updateBy(DataList<DState> datalist)
{
// Add the elements of state.vec and this->vel
for (int i = 0; i < this->data.size(); i++)  
{  
    DState temp = datalist.data[i];
    this->data[i].updateBy(temp);  
};
}

// //The explicit instantiation part
template class DataList<State>; 
template class DataList<DState>;